### PR TITLE
fix(ci): wait for k3s apiserver readiness before helm install on crio

### PR DESCRIPTION
**Pull request by Athang69**
**Purpose of PR?**: Fixes the `crio` matrix job in `ci-test-ginkgo` which fails with `apiserver not ready` because the k3s control plane is not fully stable before subsequent steps run. Confirmed failing on upstream PRs.

**Fixes**: #2622

**Does this PR introduce a breaking change?** No.

**If the changes in this PR are manually verified, list down the scenarios covered?**: The `crio` matrix job `Auto-testing Framework / oracle-vm-16cpu-64gb-x86-64 / crio` was observed failing consistently across multiple PRs. The `deploy pre existing pod` step fails with `connection to server 127.0.0.1:6443 was refused` because the apiserver is not stable after `install-k3s.sh` exits.

`Setup a Kubernetes environment` runs for 15m 51s with the wait loop spinning on connection refused:
Screenshot from 2026-05-22 16-42-32

The job then fails at `deploy pre existing pod` with apiserver still not ready:
Screenshot from 2026-05-22 16-44-23

**Additional information for reviewer?**: The `install-k3s.sh` script exits as soon as the k3s wait loop condition is met, but the apiserver is not always fully stable at that point. Subsequent steps like `deploy pre existing pod` immediately issue `kubectl apply` and `kubectl get pods` while the apiserver is still settling, causing `connection refused` errors. The fix moves two explicit readiness gates into a dedicated step right after `Setup a Kubernetes environment` and before any kubectl usage: one for node Ready condition and one for CoreDNS rollout status, which is a reliable proxy for full control plane stabilization. Previously these gates were placed inside `Run KubeArmor` which was too late. This is the same class of crio timing issue addressed in commits `35dca991` and `8b414d5b`. Failing runs observed in PRs #2539 #2527 #2531 #2530 #2538.

**Screenshot of current CI test results**: 
image

**Checklist:**
- [x] Bug fix. Fixes #
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: .github/workflows/ci-test-ginkgo.yml [hunk 1]
```diff
@@ -64,6 +64,43 @@ jobs:
...
+          for i in $(seq 1 60); do
+            if [ "$(kubectl get nodes --no-headers 2>/dev/null | wc -l)" -gt 0 ]; then
+              break
+            fi
+            sleep 5
+            if [ "$i" -eq 60 ]; then
+              echo "Timed out waiting for any Kubernetes node to be registered"
+              exit 1
+            fi
+          done
```
[the remark this answers is not in the document]
Comment (Athang69) [reply .github/workflows/ci-test-ginkgo.yml:89]: Acknowledged. These are edge case improvements to the diagnostics helper. The core fix (gating kubectl usage until nodes and CoreDNS exist) is in place. Will address diagnostics robustness in a follow-up if needed.

File: .github/workflows/ci-test-ginkgo.yml [hunk 2]
```diff
@@ -64,6 +64,43 @@ jobs:
...
+          set -Eeuo pipefail
+          dump_k3s_diagnostics() {
+            echo "::group::k3s diagnostics"
+            echo "Wait for k3s readiness failed; dumping cluster state..."
+            kubectl get nodes -o wide || true
+            kubectl -n kube-system get pods -o wide || true
+            kubectl get events -A --sort-by=.metadata.creationTimestamp | tail -n 100 || true
+            echo "::endgroup::"
+          }
+          trap 'dump_k3s_diagnostics' ERR
```
[the remark this answers is not in the document]
Comment (Athang69) [reply .github/workflows/ci-test-ginkgo.yml:78]: Acknowledged. These are edge case improvements to the diagnostics helper. The core fix (gating kubectl usage until nodes and CoreDNS exist) is in place. Will address diagnostics robustness in a follow-up if needed.

**Comment by Athang69** [comment]
@rksharma95 @Aryan-sharma11 PTAL at this PR. It fixes the intermittent `crio` matrix job failure where `deploy pre existing pod` was failing with `connection to server 127.0.0.1:6443 was refused` because the k3s apiserver was not fully stable after `install-k3s.sh` exited. The fix adds a dedicated readiness step after `Setup a Kubernetes environment` that waits for node Ready condition and CoreDNS rollout before any kubectl usage. The `blockposture` failure on the `containerd` job is a pre-existing flaky test unrelated to this PR.

**Comment by Athang69** [comment]
> Closing this as no repeated failures have been observed in recent workflow runs. 

 _Originally posted by @AryanBakliwal in [#2622](https://github.com/kubearmor/KubeArmor/issues/2622#issuecomment-4628912795)_
